substrate/runtime/treasury/src/lib.rs

```diff
@@ -0,0 +1,556 @@
+
+		// Number of proposals that have been made.
+		ProposalCount get(proposal_count): default ProposalIndex;
+
+		// Proposals that have been made.
+		Proposals get(proposals): map [ ProposalIndex => Proposal<T::AccountId, T::Balance> ];
+
+		// Proposal indices that have been approved but not yet awarded.
+		Approvals get(approvals): default Vec<ProposalIndex>;
+	}
+}
+
+/// Exported Event type that's generic over the configuration trait.
+pub type Event<T> = RawEvent<
+	<T as balances::Trait>::Balance,
+	<T as system::Trait>::AccountId,
+>;
+
+/// An event in this module.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, PartialEq, Eq, Clone)]
+pub enum RawEvent<Balance, AccountId> {
+	/// New proposal.
+	Proposed(ProposalIndex),
+	/// We have ended a spend period and will now allocate funds.
+	Spending(Balance),
+	/// Some funds have been allocated.
+	Awarded(ProposalIndex, Balance, AccountId),
+	/// Some of our funds have been burnt.
+	Burnt(Balance),
+	/// Spending has finished; this is the amount that rolls over until next spend.
+	Rollover(Balance),
+}
+
+impl<B, A> From<RawEvent<B, A>> for () {
+	fn from(_: RawEvent<B, A>) -> () { () }
+}
+
+impl<T: Trait> Module<T> {
+	/// Deposit one of this module's events.
+	fn deposit_event(event: Event<T>) {
+		<system::Module<T>>::deposit_event(<T as Trait>::Event::from(event).into());
+	}
+
+	// Implement Calls/PrivCalls and add public immutables and private mutables.
+
+	fn propose_spend(aux: &T::PublicAux, value: T::Balance, beneficiary: T::AccountId) -> Result {
+		let proposer = aux.ref_into();
+
+		let bond = Self::calculate_bond(value);
+		<balances::Module<T>>::reserve(proposer, bond)
+			.map_err(|_| "Proposer's balance too low")?;
+
+		let c = Self::proposal_count();
+		<ProposalCount<T>>::put(c + 1);
+		<Proposals<T>>::insert(c, Proposal { proposer: proposer.clone(), value, beneficiary });
+
+		Self::deposit_event(RawEvent::Proposed(c));
+
+		Ok(())
+	}
+
+	fn reject_proposal(proposal_id: ProposalIndex) -> Result {
+		let proposal = <Proposals<T>>::take(proposal_id).ok_or("No proposal at that index")?;
+		
+		let value = Self::calculate_bond(proposal.value);
+		let _ = <balances::Module<T>>::slash_reserved(&proposal.proposer, value);
+
+		Ok(())
+	}
+
+	fn approve_proposal(proposal_id: ProposalIndex) -> Result {
+		ensure!(<Proposals<T>>::exists(proposal_id), "No proposal at that index");
+
+		{
+			let mut v = <Approvals<T>>::get();
+			v.push(proposal_id);
+			<Approvals<T>>::put(v);
+		}
+		//TODO gav: make work:
+		//<Approvals<T>>::mutate(|a| a.push(proposal_id));
+
+		Ok(())
+	}
+
+	fn set_pot(new_pot: T::Balance) -> Result {
+		// Put the new value into storage.
+		<Pot<T>>::put(new_pot);
+
+		// All good.
+		Ok(())
+	}
+
+	fn configure(
+		proposal_bond: Permill,
+		proposal_bond_minimum: T::Balance,
+		spend_period: T::BlockNumber,
+		burn: Permill
+	) -> Result {
+		<ProposalBond<T>>::put(proposal_bond);
```

Comment:
This is subtle: we must make sure that this isn't called if we already have any proposals, right? Does this worth a check and/or a comment?

Reply:
Good catch. Needs further work but will just make it safe for now.